File: ios/PulsePhoneKey.swiftpm/VCSECPayload.swift
import Foundation
import CryptoKit

enum VCSECPayload {
    static let serviceUUID = "00000211-B2D1-43F0-9B88-960CEBF8B91E"
    static let writeUUID = "00000212-B2D1-43F0-9B88-960CEBF8B91E"
    static let readUUID = "00000213-B2D1-43F0-9B88-960CEBF8B91E"

    static let roleOwner: UInt64 = 2
    static let formFactorIOS: UInt64 = 6
    static let signaturePresentKey: UInt64 = 2

    static func bleLocalName(vin: String) -> String {
        let digest = Insecure.SHA1.hash(data: Data(vin.uppercased().utf8))
        let hex = digest.prefix(8).map { String(format: "%02x", $0) }.joined()
        return "S\(hex)C"
    }

    static func bleNames(vin: String) -> [String] {
        let v = vin.uppercased()
        let tail = String(v.suffix(6))
        return [bleLocalName(vin: v), "Tesla \(tail)", "Tesla\(tail)"]
    }

    static func addKeyRequest(publicKeyUncompressed: Data) -> Data {
        let publicKey = field(1, bytes: publicKeyUncompressed)
        let permissionChange = field(1, bytes: publicKey) + field(4, varint: roleOwner)
        let metadata = field(1, varint: formFactorIOS)
        let whitelistOp = field(5, bytes: permissionChange) + field(6, bytes: metadata)
        let unsigned = field(16, bytes: whitelistOp)
        let signed = field(2, bytes: unsigned) + field(3, varint: signaturePresentKey)
        let envelope = field(1, bytes: signed)
        return prependLength(envelope)
    }

    private static func prependLength(_ message: Data) -> Data {
        var out = Data([UInt8(message.count >> 8), UInt8(message.count & 0xFF)])
        out.append(message)
        return out
    }

    private static func field(_ number: UInt64, varint value: UInt64) -> Data {
        tag(number, wire: 0) + encodeVarint(value)
    }

    private static func field(_ number: UInt64, bytes: Data) -> Data {
        tag(number, wire: 2) + encodeVarint(UInt64(bytes.count)) + bytes
    }

    private static func tag(_ number: UInt64, wire: UInt64) -> Data {
        encodeVarint((number << 3) | wire)
    }

    private static func encodeVarint(_ value: UInt64) -> Data {
        var v = value
        var out = Data()
        repeat {
            var byte = UInt8(v & 0x7F)
            v >>= 7
            if v != 0 { byte |= 0x80 }
            out.append(byte)
        } while v != 0
        return out
    }
}

File: android/PulsePhoneKey/app/src/main/java/com/teslapulse/phonekey/VcsecPayload.kt
package com.teslapulse.phonekey

import java.security.MessageDigest

/** Tesla VCSEC add-key-request (same as tesla-control add-key-request). */
object VcsecPayload {
    const val SERVICE_UUID = "00000211-b2d1-43f0-9b88-960cebf8b91e"
    const val WRITE_UUID = "00000212-b2d1-43f0-9b88-960cebf8b91e"
    const val READ_UUID = "00000213-b2d1-43f0-9b88-960cebf8b91e"

    private const val ROLE_OWNER = 2L
    private const val FORM_ANDROID = 7L
    private const val SIG_PRESENT_KEY = 2L

    fun bleLocalName(vin: String): String {
        val digest = MessageDigest.getInstance("SHA-1")
            .digest(vin.uppercase().toByteArray(Charsets.UTF_8))
        val hex = digest.take(8).joinToString("") { b -> "%02x".format(b) }
        return "S${hex}C"
    }

    fun bleNames(vin: String): List<String> {
        val v = vin.uppercase()
        val tail = v.takeLast(6)
        return listOf(bleLocalName(v), "Tesla $tail", "Tesla$tail")
    }

    fun addKeyRequest(publicKeyUncompressed: ByteArray): ByteArray {
        require(publicKeyUncompressed.size == 65 && publicKeyUncompressed[0] == 0x04.toByte()) {
            "Public key must be X9.62 uncompressed (65 bytes)"
        }
        val publicKey = fieldBytes(1, publicKeyUncompressed)
        val permissionChange = fieldBytes(1, publicKey) + fieldVarint(4, ROLE_OWNER)
        val metadata = fieldVarint(1, FORM_ANDROID)
        val whitelistOp = fieldBytes(5, permissionChange) + fieldBytes(6, metadata)
        val unsigned = fieldBytes(16, whitelistOp)
        val signed = fieldBytes(2, unsigned) + fieldVarint(3, SIG_PRESENT_KEY)
        val envelope = fieldBytes(1, signed)
        return prependLength(envelope)
    }

    private fun prependLength(message: ByteArray): ByteArray {
        return byteArrayOf(
            (message.size shr 8).toByte(),
            (message.size and 0xFF).toByte(),
        ) + message
    }

    private fun fieldVarint(number: Long, value: Long): ByteArray =
        tag(number, 0) + encodeVarint(value)

    private fun fieldBytes(number: Long, data: ByteArray): ByteArray =
        tag(number, 2) + encodeVarint(data.size.toLong()) + data

    private fun tag(number: Long, wire: Long): ByteArray =
        encodeVarint((number shl 3) or wire)

    private fun encodeVarint(value: Long): ByteArray {
        var v = value
        val out = ArrayList<Byte>(10)
        while (true) {
            var b = (v and 0x7F).toInt()
            v = v ushr 7
            if (v != 0L) b = b or 0x80
            out.add(b.toByte())
            if (v == 0L) break
        }
        return out.toByteArray()
    }
}
